// rclnodejs/generated/tutorial_interfaces/tutorial_interfaces__msg__Num.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const NumRefStruct = StructType({
  lat: primitiveTypes.float64,
  lon: primitiveTypes.float64,
  alt: primitiveTypes.float64,
  dir: primitiveTypes.float64,
  velo: primitiveTypes.float64,
  x: primitiveTypes.float64,
  y: primitiveTypes.float64,
  dis: primitiveTypes.float64,
});
const NumRefArray = ArrayType(NumRefStruct);
const NumRefStructArray = StructType({
  data: NumRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NumWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NumRefStruct(other._refObject.toObject());
      this._latIntialized = true;
      this._lonIntialized = true;
      this._altIntialized = true;
      this._dirIntialized = true;
      this._veloIntialized = true;
      this._xIntialized = true;
      this._yIntialized = true;
      this._disIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NumRefStruct();
    this._latIntialized = false;
    this._lonIntialized = false;
    this._altIntialized = false;
    this._dirIntialized = false;
    this._veloIntialized = false;
    this._xIntialized = false;
    this._yIntialized = false;
    this._disIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new NumWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NumArrayWrapper;
  }
  static get ArrayType() {
    return NumArrayWrapper;
  }
  static get refObjectArrayType() {
    return NumRefStructArray
  }
  static get refObjectType() {
    return NumRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._latIntialized) {
        throw new TypeError('Invalid argument: lat in Num');
      }
      if (!this._lonIntialized) {
        throw new TypeError('Invalid argument: lon in Num');
      }
      if (!this._altIntialized) {
        throw new TypeError('Invalid argument: alt in Num');
      }
      if (!this._dirIntialized) {
        throw new TypeError('Invalid argument: dir in Num');
      }
      if (!this._veloIntialized) {
        throw new TypeError('Invalid argument: velo in Num');
      }
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in Num');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in Num');
      }
      if (!this._disIntialized) {
        throw new TypeError('Invalid argument: dis in Num');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._latIntialized = true;
    this._refObject.lat = refObject.lat;
    this._lonIntialized = true;
    this._refObject.lon = refObject.lon;
    this._altIntialized = true;
    this._refObject.alt = refObject.alt;
    this._dirIntialized = true;
    this._refObject.dir = refObject.dir;
    this._veloIntialized = true;
    this._refObject.velo = refObject.velo;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._disIntialized = true;
    this._refObject.dis = refObject.dis;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    NumWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'tutorial_interfaces', subFolder: 'msg', interfaceName: 'Num'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get lat() {
    if (!this._latIntialized) {
      return undefined;
    }
    return this._refObject.lat;
  }
  set lat(value) {
    this._latIntialized = true;
    this._refObject.lat = value;
  }
  get lon() {
    if (!this._lonIntialized) {
      return undefined;
    }
    return this._refObject.lon;
  }
  set lon(value) {
    this._lonIntialized = true;
    this._refObject.lon = value;
  }
  get alt() {
    if (!this._altIntialized) {
      return undefined;
    }
    return this._refObject.alt;
  }
  set alt(value) {
    this._altIntialized = true;
    this._refObject.alt = value;
  }
  get dir() {
    if (!this._dirIntialized) {
      return undefined;
    }
    return this._refObject.dir;
  }
  set dir(value) {
    this._dirIntialized = true;
    this._refObject.dir = value;
  }
  get velo() {
    if (!this._veloIntialized) {
      return undefined;
    }
    return this._refObject.velo;
  }
  set velo(value) {
    this._veloIntialized = true;
    this._refObject.velo = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get dis() {
    if (!this._disIntialized) {
      return undefined;
    }
    return this._refObject.dis;
  }
  set dis(value) {
    this._disIntialized = true;
    this._refObject.dis = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NumRefStruct(refObject.toObject());
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._dirIntialized = true;
    this._veloIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._disIntialized = true;
  }
  copy(other) {
    this._refObject = new NumRefStruct(other._refObject.toObject());
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._dirIntialized = true;
    this._veloIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._disIntialized = true;
  }
  static get classType() {
    return NumWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"lat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lon","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dir","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velo","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dis","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"tutorial_interfaces","type":"Num","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Num"};
  }
  hasMember(name) {
    let memberNames = ["lat","lon","alt","dir","velo","x","y","dis"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NumArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NumWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NumWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NumArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NumArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NumRefArray(size);
    this._refObject = new NumRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NumWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NumArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NumArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NumWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NumWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NumArrayWrapper;
  }
}
module.exports = NumWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "lat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lon",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dir",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velo",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dis",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "tutorial_interfaces",
    "type": "Num",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Num"
}
*/
